### DISCUSSION TITLE
Test results porting libgit2 version 1.0.1 to AIX

### DISCUSSION THREAD

**Discussion by frmichael** [libgit2/libgit2#5694] [Q&A]
Bonjour,

Porting libgit2 version 1.0.1 to AIX, we have test errors which seem to have origin in the difference Linux/AIX concerning fsync().
On AIX the file descriptor must be open with write permission for fsync() to suceed and the Posix / UNIX standard allows this behaviour.

Are these test results an indication that git/libgit2 functionality will be impared if fsync() on files opened without write permission fails  ?

The test error messages are :

 5) Failure:
odb::loose::fsync_obeys_odb_option [/opt/freeware/src/packages/BUILD/libgit2-1.0.1/64bit/tests/odb/loose.c:250]
  Function call failed: (git_odb_write(&oid, odb, "Test data\n", 10, GIT_OBJECT_BLOB))
  error -1 - failed to fsync directory 'test-objects/67': Bad file number

  6) Failure:
odb::loose::fsync_obeys_global_setting [/opt/freeware/src/packages/BUILD/libgit2-1.0.1/64bit/tests/odb/loose.c:250]
  Function call failed: (git_odb_write(&oid, odb, "Test data\n", 10, GIT_OBJECT_BLOB))
  error -1 - failed to fsync directory 'test-objects/67': Bad file number

  7) Failure:
odb::loose::fsync_obeys_repo_setting [/opt/freeware/src/packages/BUILD/libgit2-1.0.1/64bit/tests/odb/loose.c:288]
  Function call failed: (git_odb_write(&oid, odb, "Now fsync\n", 10, GIT_OBJECT_BLOB))
  error -1 - failed to fsync directory '/tmp/libgit2_tests_-sb7aa/test-objects/objects/fa': Bad file number

  8) Failure:
pack::packbuilder::fsync_global_setting [/opt/freeware/src/packages/BUILD/libgit2-1.0.1/64bit/tests/pack/packbuilder.c:218]
  Function call failed: (git_packbuilder_write(_packbuilder, ".", 0666, ((void *)0), ((void *)0)))
  error -1 - failed to fsync directory '.': Bad file number

  9) Failure:
pack::packbuilder::fsync_repo_setting [/opt/freeware/src/packages/BUILD/libgit2-1.0.1/64bit/tests/pack/packbuilder.c:227]
  Function call failed: (git_packbuilder_write(_packbuilder, ".", 0666, ((void *)0), ((void *)0)))

11) Failure:
refs::create::fsyncs_when_global_opt_set [/opt/freeware/src/packages/BUILD/libgit2-1.0.1/64bit/tests/refs/create.c:323]
  Function call failed: (git_reference_create(&ref, g_repo, "refs/heads/fsync_test", &id, 0, "log message"))
  error -1 - failed to fsync directory '/tmp/libgit2_tests_-sb7aa/testrepo/.git/logs/refs/heads': Bad file number
  12) Failure:
refs::create::fsyncs_when_repo_config_set [/opt/freeware/src/packages/BUILD/libgit2-1.0.1/64bit/tests/refs/create.c:323]
  Function call failed: (git_reference_create(&ref, g_repo, "refs/heads/fsync_test", &id, 0, "log message"))
  error -1 - failed to fsync directory '/tmp/libgit2_tests_-sb7aa/testrepo/.git/logs/refs/heads': Bad file number

**Comment by lhchavez** [comment]
To summarize the discussion from Slack, we arrived at a few conclusions:

- This is limited to [`git_futils_fsync_dir()`](https://github.com/libgit2/libgit2/blob/20450cbe7d161abbb3eb52d96867aeaa6dd99e12/src/futils.c#L1157-L1176), that `open(2)`s a *directory* (not a file) with `O_RDONLY` and uses that FD to `fsync(2)` to ensure that the metadata is flushed to disk. This is a pattern that works on some UNIX implementations (although notably not on [AIX, which requires the `fd` to be a file](https://www.ibm.com/support/knowledgecenter/ssw_aix_71/f_bostechref/fsync.html) or macOS, where a [`fcntl(2)` with `F_FULLSYNC` is preferred](https://developer.apple.com/library/archive/documentation/System/Conceptual/ManPages_iPhoneOS/man2/fsync.2.html))
- SQLite on AIX [does the same, but ignores the error from `fsync(2)`](https://www.sqlite.org/src/file?ci=tip&name=src%2Fos_unix.c&ln=3744).

**Comment by ethomson** [comment]
> SQLite on AIX does the same, but ignores the error from fsync(2).

Interesting!
